// @empirica/core/dist/chunk-SZ6OC6CP.js
// src/player/provider.ts
import { groupBy, Subject } from "rxjs";
var TajribaProvider = class {
  constructor(changes, globals, setAttributes) {
    this.globals = globals;
    this.setAttributes = setAttributes;
    this.scopes = new Subject();
    this.attributes = new Subject();
    this.participants = new Subject();
    this.steps = new Subject();
    this.dones = new Subject();
    changes.pipe(groupBy((chg) => chg.change.__typename)).subscribe({
      next: (group) => {
        switch (group.key) {
          case "ScopeChange":
            group.subscribe({
              next: (scope) => {
                this.scopes.next({
                  scope: scope.change,
                  removed: scope.removed
                });
                if (scope.done) {
                  this.dones.next();
                }
              }
            });
            break;
          case "AttributeChange":
            group.subscribe({
              next: (attribute) => {
                this.attributes.next({
                  attribute: attribute.change,
                  removed: attribute.removed
                });
                if (attribute.done) {
                  this.dones.next();
                }
              }
            });
            break;
          case "ParticipantChange":
            group.subscribe({
              next: (scope) => {
                this.participants.next({
                  participant: scope.change,
                  removed: scope.removed
                });
                if (scope.done) {
                  this.dones.next();
                }
              }
            });
            break;
          case "StepChange":
            group.subscribe({
              next: (scope) => {
                this.steps.next({
                  step: scope.change,
                  removed: scope.removed
                });
                if (scope.done) {
                  this.dones.next();
                }
              }
            });
            break;
        }
      }
    });
  }
};

// src/player/utils.ts
var isDevelopment = process.env.NODE_ENV === "development";
var isProduction = process.env.NODE_ENV === "production";
var isTest = process.env.NODE_ENV === "test";
var createNewParticipant = (key = "participantKey") => {
  var _a;
  const url = new URL(document.location.href);
  url.searchParams.set(key, new Date().getTime().toString());
  (_a = window.open(url.href, "_blank")) == null ? void 0 : _a.focus();
};

export {
  TajribaProvider,
  isDevelopment,
  isProduction,
  isTest,
  createNewParticipant
};
